tendermint-testgen/src/header.rs

```diff
@@ -0,0 +1,171 @@
+    pub time: Option<Time>,
+    #[options(help = "proposer index (default: 0)")]
+    pub proposer: Option<usize>,
+}
+
+impl Header {
+    pub fn new(validators: &[Validator]) -> Self {
+        Header {
+            validators: Some(validators.to_vec()),
+            next_validators: None,
+            chain_id: None,
+            height: None,
+            time: None,
+            proposer: None,
+        }
+    }
+    set_option!(validators, &[Validator], Some(validators.to_vec()));
+    set_option!(
+        next_validators,
+        &[Validator],
+        Some(next_validators.to_vec())
+    );
+    set_option!(chain_id, &str, Some(chain_id.to_string()));
+    set_option!(height, u64);
+    set_option!(time, Time);
+    set_option!(proposer, usize);
+}
+
+impl std::str::FromStr for Header {
+    type Err = SimpleError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let header = match parse_as::<Header>(s) {
+            Ok(input) => input,
+            Err(_) => Header::new(&parse_as::<Vec<Validator>>(s)?),
+        };
+        Ok(header)
+    }
+}
+
+impl Generator<block::Header> for Header {
+    fn merge_with_default(self, default: Self) -> Self {
+        Header {
+            validators: self.validators.or(default.validators),
+            next_validators: self.next_validators.or(default.next_validators),
+            chain_id: self.chain_id.or(default.chain_id),
+            height: self.height.or(default.height),
+            time: self.time.or(default.time),
+            proposer: self.proposer.or(default.proposer),
+        }
+    }
+
+    fn generate(&self) -> Result<block::Header, SimpleError> {
+        let vals = match &self.validators {
+            None => bail!("validator array is missing"),
+            Some(vals) => vals,
+        };
+        let vals = generate_validators(vals)?;
+        let valset = validator::Set::new(vals.clone());
+        let next_valset = match &self.next_validators {
+            Some(next_vals) => validator::Set::new(generate_validators(next_vals)?),
+            None => valset.clone(),
+        };
+        let chain_id = match chain::Id::from_str(
+            self.chain_id
+                .clone()
+                .unwrap_or("test-chain".to_string())
+                .as_str(),
+        ) {
+            Ok(id) => id,
+            Err(_) => bail!("failed to construct header's chain_id"),
+        };
+        let header = block::Header {
+            version: block::header::Version { block: 0, app: 0 },
+            chain_id,
+            height: block::Height(self.height.unwrap_or(1)),
+            time: self.time.unwrap_or(Time::now()),
+            last_block_id: None,
+            last_commit_hash: None,
+            data_hash: None,
+            validators_hash: valset.hash(),
+            next_validators_hash: next_valset.hash(),
+            consensus_hash: valset.hash(), // TODO: currently not clear how to produce a valid hash
```

Comment:
@andrey-kuprianov Could you expand on this a bit?

Reply:
well, it's that from my current limited understanding of the code base I don't see an obvious way on how to compute the consensus hash... If you know the way, please share it with me

Reply:
The consensus hash is the hash of the consensus params: https://github.com/informalsystems/tendermint-rs/blob/master/tendermint/src/consensus/params.rs

More precisely only Block.MaxBytes and Block.MaxGas are included in the message that gets hashed:
https://github.com/tendermint/tendermint/blob/940e68292cd2e710525c35df08d7e007a7c04f9d/types/params.go#L147-L166

(tmhash is just sha256)

Reply:
@liamsi thanks for the clarification! Still, seeing how it's implemented in Go, and with my current limited knowledge of Rust and how hashing is done in Tendermint, I don't feel comfortable enough to reproduce it, and be confident that what I write is correct. Also, in the current version of the light client this is not checked at all (should it, by the way?), as far as I can say.

If someone else volunteers to write this hashing part for consensus params, I am both hands for it! :) Otherwise, I would suggest to leave it for the future update.

Reply:
This is the hash of the consensus params which light clients currently don't pay attention to but which can change (though very infrequently) in the life of a blockchain. Best to open a follow up issue for things like this

Reply:
That's true; created a follow up issue #480.